enclave-runtime/src/test/mocks/handle_state_mock.rs

```diff
@@ -38,28 +43,83 @@ impl Default for HandleStateMock {
 }
 
 impl HandleState for HandleStateMock {
+	type WriteLockPayload = HashMap<ShardIdentifier, StfState>;
+
 	fn load_initialized(&self, shard: &ShardIdentifier) -> Result<StfState> {
-		self.state_map
-			.get(shard)
-			.map(|s| s.clone())
-			.ok_or_else(|| Error::Stf("No state for this shard exists".to_string()))
+		let maybe_state = self.state_map.read().unwrap().get(shard).map(|s| s.clone());
+
+		return match maybe_state {
+			// initialize with default state, if it doesn't exist yet
+			None => {
+				self.state_map.write().unwrap().insert(shard.clone(), StfState::default());
+
+				self.state_map.read().unwrap().get(shard).map(|s| s.clone()).ok_or_else(|| {
+					Error::Stf("state does not exist after inserting it".to_string())
+				})
+			},
+			Some(s) => Ok(s),
+		}
+	}
+
+	fn load_for_mutation(
+		&self,
+		shard: &ShardIdentifier,
+	) -> Result<(RwLockWriteGuard<'_, Self::WriteLockPayload>, StfState)> {
+		let initialized_state = self.load_initialized(shard)?;
+		let write_lock = self.state_map.write().unwrap();
+		Ok((write_lock, initialized_state))
 	}
 
-	fn write(&mut self, state: StfState, shard: ShardIdentifier) -> Result<H256> {
-		self.state_map.insert(shard, state);
+	fn write(
+		&self,
+		state: StfState,
+		mut state_lock: RwLockWriteGuard<'_, Self::WriteLockPayload>,
+		shard: &ShardIdentifier,
+	) -> Result<H256> {
+		state_lock.insert(shard.clone(), state);
 		Ok(H256::default())
 	}
 
 	fn exists(&self, shard: &ShardIdentifier) -> bool {
-		self.state_map.get(shard).is_some()
+		self.state_map.read().unwrap().get(shard).is_some()
 	}
 
-	fn init_shard(&mut self, shard: &ShardIdentifier) -> Result<()> {
-		self.state_map.insert(shard.clone(), StfState::new());
-		Ok(())
+	fn list_shards(&self) -> Result<Vec<ShardIdentifier>> {
+		Ok(self.state_map.read().unwrap().iter().map(|(k, _)| k.clone()).collect())
 	}
+}
 
-	fn list_shards(&self) -> Result<Vec<ShardIdentifier>> {
-		Ok(self.state_map.iter().map(|(k, _)| k.clone()).collect())
+// Since the mock itself has quite a bit of complexity, we also have tests for the mock
+#[cfg(feature = "test")]
```

Comment:
Is that not an obsolete feature, since the mocks themselves are only imported in case the `test` feature is set?